File: java/src/main/java/com/aaronicsubstances/shrewd/evolver/CustomLogEvent.java
package com.aaronicsubstances.shrewd.evolver;

import org.apache.commons.beanutils.PropertyUtils;

import com.google.gson.Gson;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;

public class CustomLogEvent {
    
    @FunctionalInterface
    public interface CustomLogEventMessageGenerator {
        String apply(Function<String, String> jsonSerializerFunc, Function<String, String> stringifyFunc);
    }

    private String message;
    private List<Object> arguments;
    private Exception error;
    private Object data;
    
    public CustomLogEvent()
    {
        this(null, null);
    }

    public CustomLogEvent(String message)
    {
        this(message, null);
    }

    public CustomLogEvent(String message, Exception error)
    {
        this.message = message;
        this.error = error;
    }

    public String getMessage() {
        return message;
    }
    public void setMessage(String message) {
        this.message = message;
    }
    
    public List<Object> getArguments() {
        return arguments;
    }
    public void setArguments(List<Object> arguments) {
        this.arguments = arguments;
    }
    
    public Exception getError() {
        return error;
    }
    public void setError(Exception error) {
        this.error = error;
    }

    public Object getData() {
        return data;
    }
    public void setData(Object data) {
        this.data = data;
    }

    public CustomLogEvent addProperty(String name, Object value) {
        if (data == null) {
            data = new LinkedHashMap<String, Object>();
        }
        ((Map<Object, Object>)data).put(name, value);
        return this;
    }
        
    public CustomLogEvent generateMessage(CustomLogEventMessageGenerator customLogEventMsgGenerator) {
        message = customLogEventMsgGenerator.apply(
            path -> fetchDataSliceAndStringify(path, true),
            path -> fetchDataSliceAndStringify(path, false));
        return this;
    }

    private String fetchDataSliceAndStringify(String path, boolean serializeAsJson) {
        Object dataSlice = fetchDataSlice(path);
        if (serializeAsJson) {
            return serializeAsJson(dataSlice);
        }
        else {
            return dataSlice != null ? dataSlice.toString() : "";
        }
    }

    Object fetchDataSlice(String path) {
        // split path and ensure no surrounding whitespace around 
        // individual segments.
        String[] pathSegments = path.split("/");
        for (int i = 0; i < pathSegments.length; i++) {
            pathSegments[i] = pathSegments[i].trim();
        }

        Object dataSlice = data;
        for (String pathSegment: pathSegments) {
            if (dataSlice == null) {
                break;
            }

            // skip empty path segments.
            if (pathSegment.length() == 0) {
                continue;
            }

            if (dataSlice instanceof List) {
                List<Object> jsonArray = (List<Object>) dataSlice;
                try {
                    int index = Integer.parseInt(pathSegment);
                    // Support Python style indexing.
                    if (Math.abs(index) >= jsonArray.size()) {
                        dataSlice = null;
                    }
                    else {
                        if (index < 0)
                        {
                            index += jsonArray.size();
                        }
                        dataSlice = jsonArray.get(index);
                    }
                    continue;
                }
                catch (NumberFormatException ignore) {
                    
                }

                // let Map branch of code handle pathSegments which are ints 
                // but correspond to data slices which
                // are not list instances.
            }

            if (dataSlice instanceof Map) {
                Map<Object, Object> dataSliceDict = (Map<Object, Object>) dataSlice;
                if (dataSliceDict.containsKey(pathSegment)) {
                    dataSlice = dataSliceDict.get(pathSegment);
                }
                else {
                    dataSlice = null;
                }
            }
            else {
                dataSlice = getTreeDataPropertyValue(dataSlice, pathSegment);
            }
        }
        return dataSlice;
    }

    protected Object getTreeDataPropertyValue(Object dataSlice, String pathSegment) {
        if (dataSlice == null) {
            return null;
        }
        try {
            // try declared property access
            return PropertyUtils.getSimpleProperty(dataSlice, pathSegment);
        }
        catch (Exception ignore) {
            return null;
        }
    }

    protected String serializeAsJson(Object o) {
        return new Gson().toJson(o);
    }
}

File: csharp/ShrewdEvolver/CustomLogEvent.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using System.Text;

namespace AaronicSubstances.ShrewdEvolver
{
    public class CustomLogEvent
    {
        public delegate string CustomLogEventMessageGenerator(
            Func<string, string> jsonSerializerFunc, Func<string, string> stringifyFunc);

        public CustomLogEvent(Type targetLogger)
            : this(targetLogger, null, null)
        { }

        public CustomLogEvent(Type targetLogger, string message)
            : this(targetLogger, message, null)
        { }

        public CustomLogEvent(Type targetLogger, string message, Exception error)
        {
            Message = message;
            Error = error;
            TargetLogger = targetLogger?.FullName;
        }

        public string Message { get; set; }
        public List<object> Arguments { get; set; }
        public Exception Error { get; set; }
        public object Data { get; set; }
        public string TargetLogger { get; set; }

        public CustomLogEvent AddProperty(string name, object value)
        {
            if (Data == null)
            {
                Data = new Dictionary<string, object>();
            }
            ((IDictionary<string, object>)Data).Add(name, value);
            return this;
        }
        
        public CustomLogEvent GenerateMessage(CustomLogEventMessageGenerator customLogEventMsgGenerator)
        {
            Message = customLogEventMsgGenerator.Invoke(
                path => FetchDataSliceAndStringify(path, true),
                path => FetchDataSliceAndStringify(path, false));
            return this;
        }

        private string FetchDataSliceAndStringify(string path, bool serializeAsJson)
        {
            object dataSlice = FetchDataSlice(path);
            if (serializeAsJson)
            {
                return SerializeAsJson(dataSlice);
            }
            else
            {
                return dataSlice?.ToString() ?? "";
            }
        }

        internal object FetchDataSlice(string path)
        {
            // split path and ensure no surrounding whitespace around 
            // individual segments.
            var pathSegments = path.Split('/');
            for (int i = 0; i < pathSegments.Length; i++)
            {
                pathSegments[i] = pathSegments[i].Trim();
            }

            object dataSlice = Data;
            foreach (var pathSegment in pathSegments)
            {
                if (dataSlice == null)
                {
                    break;
                }

                // skip empty path segments.
                if (pathSegment.Length == 0)
                {
                    continue;
                }

                if (dataSlice is IList<object> jsonArray)
                {
                    if (int.TryParse(pathSegment, out int index))
                    {
                        // Support Python style indexing.
                        if (Math.Abs(index) >= jsonArray.Count)
                        {
                            dataSlice = null;
                        }
                        else
                        {
                            if (index < 0)
                            {
                                index += jsonArray.Count;
                            }
                            dataSlice = jsonArray[index];
                        }
                        continue;
                    }

                    // let dictionary branch of code handle pathSegments which are ints 
                    // but correspond to data slices which
                    // are not list instances.
                }

                if (dataSlice is IDictionary<string, object> dataSliceDict)
                {
                    if (dataSliceDict.ContainsKey(pathSegment))
                    {
                        dataSlice = dataSliceDict[pathSegment];
                    }
                    else
                    {
                        dataSlice = null;
                    }
                }
                else
                {
                    dataSlice = GetTreeDataPropertyValue(dataSlice, pathSegment);
                }
            }
            return dataSlice;
        }

        protected virtual object GetTreeDataPropertyValue(object dataSlice, string pathSegment)
        {
            if (dataSlice == null)
            {
                return null;
            }
            try
            {
                // try dictionary access. Probably IDictionary<string, some_type_not_object>
                dynamic dataSliceDict = dataSlice;
                return dataSliceDict[pathSegment];
            }
            catch (Exception)
            {
                try
                {
                    // try declared property access
                    var propHandle = dataSlice.GetType().GetProperty(pathSegment);
                    return propHandle.GetValue(dataSlice);
                }
                catch (Exception)
                {
                    return null;
                }
            }
        }

        protected virtual string SerializeAsJson(object o)
        {
            return JsonConvert.SerializeObject(o);
        }
    }
}
